### PR TITLE
Introducing new event for failed http requests #106284

### LINKED ISSUE #106284: Add RequestFailed event to default HttpClientFactory logging
## Suggestion

I'd like to add logging for failed http requests.

## Motivation

Current [LoggingHttpMessageHandler](https://github.com/dotnet/runtime/blob/c5a3f49c88d3d907a56ec8d18f783426de5144e9/src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs) logs requests/responses and headers and everything is fine.

If a request fails `LoggingHttpMessageHandler` doesn't log any failed message. It is not an issue for regular use cases, but if you collect logs for analyzing it can be a challenge to correlate requests and failed messages.  

Introducing the feature will make all logs more consistence and should not affect any performance. 

## Action plan

1. Introduce new EventId (id: 104, name: "RequestFailed")
2. Introduce a model for the event (TimeSpan Duration, int StatusCode, string ErrorMessage)
3. Wrap http call method with `try/catch`
4. in `catch` section log the failed request

### API Proposal

```csharp
// src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs
namespace Microsoft.Extensions.Http.Logging;

public class LoggingHttpMessageHandler : DelegatingHandler
{
    /* existing code */
   
   protected async override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            /* existing code */

            // Not using a scope here because we always expect this to be at the end of the pipeline, thus there's
            // not really anything to surround.
            Log.RequestStart(_logger, request, shouldRedactHeaderValue);
            try
            {
               HttpResponseMessage response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
               Log.RequestEnd(_logger, response, stopwatch.GetElapsedTime(), shouldRedactHeaderValue);
               return response;
            } catch(HttpRequestException ex)
            {
               Log.RequestEnd(_logger, stopwatch.GetElapsedTime(), ex.Message);
               throw;
            }
        }

   internal static class Log 
   {
      public static class EventIds 
      {
         /* existing code */
         public static readonly EventId RequestFailed = new EventId(105, "RequestFailed");
          /* existing code */
      }
     
      private static readonly Action<ILogger, double, int, Exception> _requestFailed = LoggerMessage.Define<double, string>(
         LogLevel.Error,
         EventIds.RequestFailed ,
         "HTTP request fails after {ElapsedMilliseconds}ms. Error message: {ErrorMessage}");

      public static void RequestEnd(ILogger logger, TimeSpan duration, string errorMessage)
      {
         _requestFailed(logger, duration.TotalMilliseconds, errorMessage);
      }
   }
}
```

### API Usage

```csharp
//
public static MyDTO? ParseRequestPipelineRequestHeader(TraceEvent traceEvent)
{
   var eventId = Convert.ToInt32(value.PayloadByName("EventId"));

   if(eventId != 105)
   {
      return null;
   }

    var headersAsText = traceEvent.PayloadByName("FormattedMessage")?.ToString();

    var result = ParseHeaders(headersAsText);

   var elapsedMilliseconds = arguments["ElapsedMilliseconds"];
   var errorMessage = arguments["ErrorMessage"];   

   // will be used to corelate with a request;
    var correlationId = traceEvent.ActivityID.ToString();

    return new MyDTO 
    {
        CorrelationId = correlationId,
        ElapsedMilliseconds  = elapsedMilliseconds ,
        ErrorMessage  = errorMessage
    };
}
```

### Alternative Designs

As an alternative design can be a solution to integrate failed log fields to RequestEnd event.

### Risks

I see no risks here, but please feel free to share your thoughs

### PR DESCRIPTION
**Pull request by OlzhabaevSh**
Please visit #106284 to learn more.

## Highlights

* Added new event Id `103`
* Failed requests are wrapped by try/catch blocks
* `HttpRequestException ` is available from dotnet 8 and higher. for the other version I return `500` status code.

---

Fixes https://github.com/dotnet/runtime/issues/106284

### CODE REVIEW

**Comment by OlzhabaevSh** [comment]
@dotnet-policy-service agree

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs [hunk 1]
```diff
@@ -59,16 +59,31 @@ async Task<HttpResponseMessage> Core(HttpRequestMessage request, bool useAsync,
...
-                return response;
+                    return response;
+                }
+                catch (HttpRequestException ex)
+                {
+                    var statusCode =
+#if NET_8_OR_GREATER
+                        (int)ex.StatusCode;
+#else
+                        500;
```
[the remark this answers is not in the document]
Comment (OlzhabaevSh) [reply src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:82]: Hey @huoyaoyuan,
Thank you for the comment!
Yes, you are right:
In case if a request completes and returns failure status, it means that we will have a response log. 

The case what I'm trying to solve is about when a request fails (because if network things) and we can't expect any status codes there. 

I'll need to update the PR.

**Comment by OlzhabaevSh** [comment]
Hey @huoyaoyuan 

The runtime pipelines fails and as far as I can see it is because of too long execution. 

I'm not really sure how the changes can be the reason of it...

**Comment by rzikm** [comment]
> The runtime pipelines fails and as far as I can see it is because of too long execution.

Those failures sometimes happen, they are not necessarily caused by your changes. I suggest you ignore these now while you while you work on applying the code review feedback.

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (huoyaoyuan) [reply src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:82]: Then the event definition also need to be reconsidered. Status code isn't really applicable in this case.

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs [hunk 2]
```diff
@@ -112,6 +127,11 @@ public static void LogRequestEnd(this ILogger logger, HttpResponseMessage respon
             }
         }
 
+        public static void LogRequestFailed(this ILogger logger, int statusCode, TimeSpan duration, string errorMessage)
+        {
+            _requestFailed(logger, duration.TotalMilliseconds, statusCode, errorMessage, null);
```
Comment (MihaZupan) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:132]: We should pass the exception object to the logger to avoid losing useful information.
The exception message itself is often not useful (cancelled vs failed) compared to the stack trace and inner exceptions.

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (OlzhabaevSh) [reply src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs]: Hey @huoyaoyuan!
After some investigations I decided to keep only the duration and error message.

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs (hunk 2 above)
Comment (OlzhabaevSh) [reply src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs]: Hey @MihaZupan ,
Updated

**Comment by OlzhabaevSh** [comment]
@MihaZupan @huoyaoyuan 
I just noticed that there is a class [HttpClientLoggerHandler](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Http/src/Logging/HttpClientLoggerHandler.cs) which is not used and not even fully completed, but I'm wondering don't we just repeating the same thing? 

I'm curious if that class is going to be used in near feature?

**Comment by CarnaViire** [comment]
Sorry for the delay @OlzhabaevSh, I was on vacation. I will review the changes shortly.

> a class [HttpClientLoggerHandler](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Http/src/Logging/HttpClientLoggerHandler.cs) which is not used and not even fully completed

I'm not sure what you mean by "not fully completed"?

This handler is used when the end-user specifies their own logging implementation ([IHttpClientLogger](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.http.logging.ihttpclientlogger?view=net-8.0) implementation) with the [AddLogger](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.httpclientbuilderextensions.addlogger?view=net-8.0) API. In the code, you can see it being used [here](https://github.com/dotnet/runtime/blob/34d2f179b6e5886bab33cd9e6882eebc48bd6b5b/src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.Logging.cs#L56).

> I'm curious if that class is going to be used in near feature?

It is already used; but it is true, however, that the legacy default logging handlers (`LoggingHttpMessageHandler` and `LoggingScopeHttpMessageHandler`) are not leveraging this new approach at the moment. It would be quite convenient to finally unify them, but it's just a nice-to-have, so it's currently on a back burner.

Let me know if you might be interested in helping with this refactoring as well (it can but doesn't have to be within this PR)?

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs [hunk 3]
```diff
@@ -45,6 +49,12 @@ private static class EventIds
             EventIds.RequestEnd,
             "Received HTTP response headers after {ElapsedMilliseconds}ms - {StatusCode}");
 
+        private static readonly Action<ILogger, double, string?, Exception?> _requestFailed = LoggerMessage.Define<double, string?>(
+            LogLevel.Error,
+            EventIds.RequestFailed,
+            "Request failed after {ElapsedMilliseconds}ms, Error: {ErrorMessage}");
```
Comment (CarnaViire) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:55]: 1) Let's try to keep all the messages in the same style within the logging handler. The request is called "HTTP request" in other messages: `"HTTP request failed after {ElapsedMilliseconds}ms"`

2) I believe you don't need to additionally pass the ErrorMessage here. You are already passing the `Exception` object, so it will already be logged for you. E.g. you can check the [Logging in C# - Log exceptions](https://learn.microsoft.com/en-us/dotnet/core/extensions/logging?tabs=command-line#log-exceptions) section, and the example in the [High-performance logging - Define a logger message](https://learn.microsoft.com/en-us/dotnet/core/extensions/high-performance-logging#define-a-logger-message) section.

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs [hunk 4]
```diff
@@ -57,6 +67,11 @@ private static class EventIds
             EventIds.PipelineEnd,
             "End processing HTTP request after {ElapsedMilliseconds}ms - {StatusCode}");
 
+        private static readonly Action<ILogger, double, string?, Exception?> _requestPipelineFailed = LoggerMessage.Define<double, string?>(
+            LogLevel.Error,
+            EventIds.PipelineFailed,
+            "Request failed after {ElapsedMilliseconds}ms, Error: {ErrorMessage}");
```
Comment (CarnaViire) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:73]: Same here with regards to the unified wording and the ErrorMessage parameter: `"Error while processing HTTP request after {ElapsedMilliseconds}ms"`

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs [hunk 5]
```diff
@@ -25,11 +25,15 @@ private static class EventIds
...
 
+            public static readonly EventId RequestFailed = new EventId(104, "RequestFailed");
+
             public static readonly EventId PipelineStart = new EventId(100, "RequestPipelineStart");
             public static readonly EventId PipelineEnd = new EventId(101, "RequestPipelineEnd");
 
             public static readonly EventId RequestPipelineRequestHeader = new EventId(102, "RequestPipelineRequestHeader");
             public static readonly EventId RequestPipelineResponseHeader = new EventId(103, "RequestPipelineResponseHeader");
+
+            public static readonly EventId PipelineFailed = new EventId(104, "RequestPipelineRequestFailed");
```
Comment (CarnaViire) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:36]: To align with PipelineStart and PipelineEnd event IDs:
```suggestion
            public static readonly EventId PipelineFailed = new EventId(104, "RequestPipelineFailed");
```

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs [hunk 6]
```diff
@@ -59,16 +59,25 @@ async Task<HttpResponseMessage> Core(HttpRequestMessage request, bool useAsync,
                 using (_logger.BeginRequestPipelineScope(request, out string? formattedUri))
                 {
                     _logger.LogRequestPipelineStart(request, formattedUri, shouldRedactHeaderValue);
-                    HttpResponseMessage response = useAsync
+
+                    try
+                    {
+                        HttpResponseMessage response = useAsync
```
Comment (CarnaViire) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs:65]: Please fix the formatting in the block (LoggingHttpMessageHandler above seems to be correct)

**Review by CarnaViire — CHANGES_REQUESTED** [review]
Please also a test to verify the new logs are emitted. You can get inspiration from [RedactedLogValueIntegrationTest](https://github.com/dotnet/runtime/blob/ad0c47c42e8901b9300701b237c42345b83c2bd9/src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/Logging/RedactedLogValueIntegrationTest.cs)

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs [hunk 7]
```diff
@@ -45,6 +49,12 @@ private static class EventIds
             EventIds.RequestEnd,
             "Received HTTP response headers after {ElapsedMilliseconds}ms - {StatusCode}");
 
+        private static readonly Action<ILogger, double, string?, Exception?> _requestFailed = LoggerMessage.Define<double, string?>(
+            LogLevel.Error,
```
[the remark this answers is not in the document]
Comment (OlzhabaevSh) [reply src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs]: @CarnaViire  done

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs (hunk 3 above)
Comment (OlzhabaevSh) [reply src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:55]: @CarnaViire done

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs (hunk 4 above)
Comment (OlzhabaevSh) [reply src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:73]: @CarnaViire done

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs (hunk 5 above)
Comment (OlzhabaevSh) [reply src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs]: @CarnaViire done

**Comment by OlzhabaevSh** [comment]
> Sorry for the delay @OlzhabaevSh, I was on vacation. I will review the changes shortly.
> 
> > a class [HttpClientLoggerHandler](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Http/src/Logging/HttpClientLoggerHandler.cs) which is not used and not even fully completed
> 
> I'm not sure what you mean by "not fully completed"?
> 
> This handler is used when the end-user specifies their own logging implementation ([IHttpClientLogger](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.http.logging.ihttpclientlogger?view=net-8.0) implementation) with the [AddLogger](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.httpclientbuilderextensions.addlogger?view=net-8.0) API. In the code, you can see it being used [here](https://github.com/dotnet/runtime/blob/34d2f179b6e5886bab33cd9e6882eebc48bd6b5b/src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.Logging.cs#L56).
> 
> > I'm curious if that class is going to be used in near feature?
> 
> It is already used; but it is true, however, that the legacy default logging handlers (`LoggingHttpMessageHandler` and `LoggingScopeHttpMessageHandler`) are not leveraging this new approach at the moment. It would be quite convenient to finally unify them, but it's just a nice-to-have, so it's currently on a back burner.
> 
> Let me know if you might be interested in helping with this refactoring as well (it can but doesn't have to be within this PR)?

Oh I see... yep, you are right. the class is used. My bad.

If you have any information or expectation how the unification should look like, I can try to make it :-)

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs (hunk 3 above)
Comment (CarnaViire) [reply src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:55]: The message is not updated -- can you pls check (maybe the changes were not committed/pushed?)

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs (hunk 4 above)
Comment (CarnaViire) [reply src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:73]: The message is not updated -- can you pls check (maybe the changes were not committed/pushed?)

**Comment by CarnaViire** [comment]
@OlzhabaevSh Please also a test to verify the new logs are emitted. You can get inspiration from [RedactedLogValueIntegrationTest](https://github.com/dotnet/runtime/blob/ad0c47c42e8901b9300701b237c42345b83c2bd9/src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/Logging/RedactedLogValueIntegrationTest.cs)

**Comment by OlzhabaevSh** [comment]
> @OlzhabaevSh Please also a test to verify the new logs are emitted. You can get inspiration from [RedactedLogValueIntegrationTest](https://github.com/dotnet/runtime/blob/ad0c47c42e8901b9300701b237c42345b83c2bd9/src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/Logging/RedactedLogValueIntegrationTest.cs)

Sure, I introduced a new test `FailedHttpRequestLoggingTests` for this purpose.

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs [hunk 8]
```diff
@@ -45,6 +49,12 @@ private static class EventIds
             EventIds.RequestEnd,
             "Received HTTP response headers after {ElapsedMilliseconds}ms - {StatusCode}");
 
+        private static readonly Action<ILogger, double, Exception?> _requestFailed = LoggerMessage.Define<double>(
+            LogLevel.Information,
+            EventIds.RequestFailed,
+            "HTTP request failed after {ElapsedMilliseconds}ms");
+
+
```
Comment (CarnaViire) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:57]: Duplicate empty line

```suggestion

```

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs [hunk 9]
```diff
@@ -112,6 +127,11 @@ public static void LogRequestEnd(this ILogger logger, HttpResponseMessage respon
             }
         }
 
+        public static void LogRequestFailed(this ILogger logger, TimeSpan duration, HttpRequestException exception)
+        {
+            _requestFailed(logger, duration.TotalMilliseconds, exception);
+        }
```
Comment (CarnaViire) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:133]: Nit:

```suggestion
        public static void LogRequestFailed(this ILogger logger, TimeSpan duration, HttpRequestException exception)
            => _requestFailed(logger, duration.TotalMilliseconds, exception);
```

File: src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs [hunk 10]
```diff
@@ -148,6 +168,11 @@ public static void LogRequestPipelineEnd(this ILogger logger, HttpResponseMessag
             }
         }
 
+        public static void LogRequestPipelineFailed(this ILogger logger, TimeSpan duration, HttpRequestException exception)
+        {
+            _requestPipelineFailed(logger, duration.TotalMilliseconds, exception);
+        }
```
Comment (CarnaViire) [inline src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:174]: Nit:

```suggestion
        public static void LogRequestPipelineFailed(this ILogger logger, TimeSpan duration, HttpRequestException exception)
            => _requestPipelineFailed(logger, duration.TotalMilliseconds, exception);
```

File: src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/Logging/FailedHttpRequestLoggingTests.cs [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+        private static class EventIds
+        {
+            public static readonly EventId RequestHeader = new EventId(102, "RequestHeader");
+            public static readonly EventId ResponseHeader = new EventId(103, "ResponseHeader");
+            public static readonly EventId RequestFailed = new EventId(104, "RequestFailed");
+
+            public static readonly EventId RequestPipelineRequestHeader = new EventId(102, "RequestPipelineRequestHeader");
+            public static readonly EventId RequestPipelineResponseHeader = new EventId(103, "RequestPipelineResponseHeader");
+            public static readonly EventId PipelineFailed = new EventId(104, "RequestPipelineFailed");
```
Comment (CarnaViire) [inline src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/Logging/FailedHttpRequestLoggingTests.cs:30]: I believe other ids are not used in this test

```suggestion
            public static readonly EventId RequestFailed = new EventId(104, "RequestFailed");
            public static readonly EventId PipelineFailed = new EventId(104, "RequestPipelineFailed");
```

File: src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/Logging/FailedHttpRequestLoggingTests.cs [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+            serviceCollection.AddSingleton<ILoggerFactory>(new TestLoggerFactory(sink, enabled: true));
+
+            // Act
+            serviceCollection
+                .AddHttpClient("test")
+                .ConfigurePrimaryHttpMessageHandler(() => new FailMessageHandler())
+                .RedactLoggedHeaders(header =>
+                {
+                    return header.StartsWith("Auth") || header.StartsWith("X-");
+                });
```
Comment (CarnaViire) [inline src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/Logging/FailedHttpRequestLoggingTests.cs:50]: The header redaction also doesn't seem relevant to the test, so I think we can remove it from setup
```suggestion
                .ConfigurePrimaryHttpMessageHandler(() => new FailMessageHandler());
```

**Review by CarnaViire — APPROVED** [review]
Minor (mostly formatting) nits, otherwise looks good to me, thanks a lot!

Pls address the nits, then after CI passes it will be ready to merge :shipit:
